### DISCUSSION TITLE
Return Types of Functions

### DISCUSSION THREAD

**Discussion by DerPicknicker** [CarlosDerSeher/snapclient#51] [General]
Hi @CarlosDerSeher ,

I am wondering how the code is structured. As you know I am working on a new DAC-Type (TAS5805M). I got some information how to use the dac and luckily the implementation is almost finished.

For me it's not clear how you use the functions. For example here: 

`esp_err_t adau1961_get_volume(int *vol)`

Why you use the return type esp_err_t and not an int instead? 

Did I miss something? - Is the **esp_err_t** the preferred method to use in this case?

Thank you in advance!

**Comment by CarlosDerSeher** [comment]
You need some kind of error checking inside this function. I2C communication could fail for some reason and this needs to be returned to the calling function somehow. Thats why volume is returned by reference through the parameter.

esp_err_t is in fact an int I belive. I was always focused on IDF and they use this type.

**Comment by DerPicknicker** [comment]
Hello again,

thank you for the explanation. I will try that. At some point I got lost...Now I can move on. I never used that way of returning a value. For me it was easier to use an explicit type (Boolean, String or whatever). 

If I have finished all functions how should I test the code. Do you have an recommended way?

**Reply by CarlosDerSeher** [reply]
I am kind of the guy who takes a pragmatic approach. Test it using real hardware, mute it crank the volume up and down etc.

And these types are just typedef'ed enums most of the time. You can look at them in a header file where the are defined.

**Reply by DerPicknicker** [reply]
I will try that. The snapserver is already running.

The missing part is only the esp32. I will keep you updated.

**Reply by CarlosDerSeher** [reply]
Also if you have the possibility to measure power consumption you could test if power down options of your DAC are working as intended.

**Reply by DerPicknicker** [reply]
@CarlosDerSeher .. I can measure it. 

The main thing is coding. As an pure c programmer it's hard to get into the concept with the esp_err_t and with the reference to the parameters.

Do you have a good example how I can call a function from the type esp_err_t which should also return a Boolean? I want to make the code as clear and simple as possible.

**Reply by CarlosDerSeher** [reply]
You could always pass a parameter by reference

`esp_err_t my_func(int a, bool *b)`

then

```
bool b;
int myInt=1;

if (my_func(myInt, &b) != ESP_OK) {
  LOG("error");

  return;
}

// do something with b
```

**Reply by DerPicknicker** [reply]
Hi, 

I think it wasn't 100% clear. 

```
esp_err_t ma120_get_volume(int *vol) {
  esp_err_t ret = ESP_OK;
  uint8_t rxbuf;
  rxbuf = ma_read_byte(MA120_ADDR, 2, 3);
  *vol = 128 - rxbuf;
  return ret;
}
```

This function should return the volume, right? But where are the values are stored... I want to use the volume in my mute function because the TAS5805M is set to mute if the volume is set to 255 and if this is already the case I don't have to write again i2C commands. Which saves some cpu time. 

Edit: Your example is correct if you're using global variables but in the header files I can't find any of them. This is why I am asking.

**Reply by CarlosDerSeher** [reply]
Then you'll need to declare a static global variable in your driver's c module and save the set volume there for later usage.

```
static int currentVol = 0;

esp_err_t ma120_get_volume(int *vol) {
  esp_err_t ret = ESP_OK;
  uint8_t rxbuf;
  rxbuf = ma_read_byte(MA120_ADDR, 2, 3);
  *vol = 128 - rxbuf;
  currentVol = *vol;
  return ret;
}
```

**Comment by DerPicknicker** [comment]
Hi @CarlosDerSeher .. 

It's harder than I thought. I never used that IDF in such a low level. So it might be good if you could briefly explain for what these declarations are used: 
```
esp_err_t tas5805m_ctrl(audio_hal_codec_mode_t mode,
                        audio_hal_ctrl_t ctrl_state);
esp_err_t tas5805m_conig_iface(audio_hal_codec_mode_t mode,
                               audio_hal_codec_i2s_iface_t *iface);
```

Can I left them as they are? (I copied them from another driver..) 


EDIT2: 
Ignore below. I worked in the wrong directory. Where the ADF drivers are located. Big sorry for that. May I ask you to check the code briefly before I hook everything up (it's a bit tricky). So it would be good if a second instance checks it before. Thanks.

**EDIT: **

I got the Project building with my additions to the driver. I didn't hook my board up (until now). Can I ignore these warnings? 

```
./components/audio_hal/driver/tas5805m/tas5805m.c:68:29: warning: initialization of 'esp_err_t (*)(_Bool)' {aka 'int (*)(_Bool)'} from incompatible pointer type 'esp_err_t (*)()' {aka 'int (*)()'} [-Wincompatible-pointer-types]
     .audio_codec_set_mute = tas5805m_set_mute,
                             ^~~~~~~~~~~~~~~~~
../components/audio_hal/driver/tas5805m/tas5805m.c:68:29: note: (near initialization for 'AUDIO_CODEC_TAS5805M_DEFAULT_HANDLE.audio_codec_set_mute')
../components/audio_hal/driver/tas5805m/tas5805m.c: In function 'tas5805m_get_volume':
../components/audio_hal/driver/tas5805m/tas5805m.c:136:64: warning: passing argument 2 of 'tas5805m_read_byte' from incompatible pointer type [-Wincompatible-pointer-types]
   int ret = tas5805m_read_byte(TAS5805M_DIG_VOL_CTRL_REGISTER, &VolumeLevel);
                                                                ^~~~~~~~~~~~
../components/audio_hal/driver/tas5805m/tas5805m.c:75:62: note: expected 'uint8_t *' {aka 'unsigned char *'} but argument is of type 'uint8_t **' {aka 'unsigned char **'}
 esp_err_t tas5805m_read_byte(uint8_t register_name, uint8_t *data)
                                                     ~~~~~~~~~^~~~
```
If yes I can do the next steps (adding it to the Menu etc..)

**Reply by CarlosDerSeher** [reply]
Have a look here, [adau1961_ctrl](https://github.com/CarlosDerSeher/snapclient/blob/d3e20fe0eb4d8d562d97a02f43cbccb79e9c8505/components/custom_board/adau1961/adau1961.c#L611C11-L611C24) just leave it blank for now. You could also have a look at [lyrat driver](https://github.com/CarlosDerSeher/snapclient/blob/d3e20fe0eb4d8d562d97a02f43cbccb79e9c8505/components/audio_hal/driver/es8388/es8388.c#L602) it is implemented there, but I think you won't need it if you just want a playback device. In lyrat driver you can see what these functions [should do](https://github.com/CarlosDerSeher/snapclient/blob/d3e20fe0eb4d8d562d97a02f43cbccb79e9c8505/components/audio_hal/driver/es8388/es8388.c#L47)

Yes I can check, did you push to your fork already?

**Reply by DerPicknicker** [reply]
I created a new branch. From your current master. I will fork it and push it. I will write here again if I am ready.

**Reply by DerPicknicker** [reply]
See here: 

https://github.com/CarlosDerSeher/snapclient/pull/52
